core-to-plc/src/Language/Plutus/CoreToPLC.hs

```diff
@@ -317,6 +363,79 @@ type of the overall case expression.
 - We apply the instantiated scrutinee to the functions for each alternative.
 -}
 
+{- Note [Abstract data types]
+While the Scott encoding makes it easy to simply convert all types inline, it
+is convenient for a number of reasons to instead abstract out data types. Namely:
+- The resulting code is much more readable, since we have named types instead
+of (potentially big) inlined types.
+- We generate less code because we're not repeating the definition at every use site.
+
+The basic idea is to "let-bind" types using type abstractions, and values using
+lambda abstractions. There are a few considerations that make this tricky, however:
+1. When types are inlined, the Scott encoding allows us to construct values by
+constructing the matching function inline. When they are abstract, we need to provide
+(suitably polymorphic) constructors abstractly.
+2. When types are inlined, the Scott encoding allows us to match against a type by
+simply using it as a function. When they are abstract, we need to provide a
+(suitably polymorphic) matching function abstractly.
+3. The definition of a type must be abstract in the binding for the constructors and destructors
+(so they can be used alongside the abstract type), but it must *not* be abstract in the
+*body* of the constructors or destructors, because they depend on knowing the real structure
+of the type.
+
+Consequently, for a single type we end up with something like the following:
+
+(abs ty :: (type) .
+  \c1 : forall t_1 :: (type) .. t_i :: (type) . t_1 -> .. -> t_c_1_k -> ty .. c_j : t_1 -> t_c_j_k -> ty
+    \match : forall t_1 :: (type) .. t_i :: (type) . forall r :: (type) b. (t_1 -> .. -> t_c_1_k -> r) -> .. -> (t_1 -> t_c_j_k -> r) -> r
+      <user term>
+)
+<defn of t>
+<defn of c1>
+..
+<defn of c_n>
+<defn of match>
+(see Note [Scott-encoding of datatypes] for how the actual definitions are constructed)
+
+We also have to convert data constructors and case expressions into suitably instantiated
+calls to the constructors/destructors instead of converting them directly as described in
+Note [Scott encoding of data types]
+
+The requirements imposed by 3. render translation somewhat tricky. We want our translation to be
+ambivalent about where we are in the process of creating a type, so we use a richer type
```

Comment:
I'd appreciate an explanation of why we need the ambivalence. I.e. which parts of the code work with both the visible and abstract data types?

Reply:
All of it! In particular conversion of types, where we want to handle e.g.
```
data List a = Nil | Cons a (List a) -- here `List` is the recursive variable
data ListWrapper a = ListWrapper (List a) -- here `List` is a reference to the opaque type
```
without the code that handles the types knowing that anything is different. Which works very nicely here. I'll add a comment.

Reply:
See now. Thanks.